firewood/src/merkle/changes.rs

```diff
@@ -0,0 +1,389 @@
+        let mut path = pre_path;
+        path.extend(node.partial_path().iter().copied());
+
+        Ok(Self {
+            path,
+            node,
+            hash: child.hash().map(|hash| hash.clone().into_triehash()),
+        })
+    }
+}
+
+/// Contains the state required for performing pre-order iteration of a Merkle trie. This includes
+/// the state of the current node, a reference to a trie that implements the `HashedNodeReader`
+/// trait, and a stack that contains the `NodeState` of nodes to be traversed with calls to `next`.
+struct PreOrderIterator<'a, T: HashedNodeReader> {
+    node_state: Option<NodeState>,
+    trie: &'a T,
+    traversal_stack: Vec<NodeState>,
+}
+
+impl<'a, T: HashedNodeReader> PreOrderIterator<'a, T> {
+    /// Create a pre-order iterator for the trie that starts at `start_key`.
+    #[cfg_attr(not(test), expect(dead_code))]
+    fn new(trie: &'a T, start_key: &Key) -> Result<PreOrderIterator<'a, T>, FileIoError> {
+        let mut preorder = Self {
+            node_state: None,
+            traversal_stack: vec![],
+            trie,
+        };
+        // If the root node is not None, then push a `NodeState` for the root onto the traversal
+        // stack. It will be used on the first call to `next` or `next_internal`. Because we
+        // already have the root node, we create its `NodeState` directly instead of using
+        // `NodeState::new` as we don't need to create a `SharedNode` from a `Child`. The full
+        // path of the root node is just its partial path since it has no pre-path.
+        if let Some(root) = trie.root_node() {
+            preorder.traversal_stack.push(NodeState {
+                path: root.partial_path().clone(),
+                node: root,
+                hash: trie.root_hash().map(|hash| hash.clone().into_triehash()),
+            });
+        }
+        preorder.iterate_to_key(start_key)?;
+        Ok(preorder)
+    }
+
+    /// In a textbook implementation of pre-order traversal we would normally pop off the next node
+    /// from the traversal stack and push its children onto the stack before returning. However, that
+    /// would be inefficient if we want to skip traversing a node's children if its hash matches that
+    /// of a node in a different trie, as we would then need to pop its children off the traversal
+    /// stack.
+    ///
+    /// This implementation flips the order such that we don't push a node's children onto the
+    /// traversal stack until the subseqent call to `next`. This is done by saving the node state of
+    /// the current node in `node_state` and keeping that available for the next call. Skipping
+    /// traversal of the children then just involves setting `node_state` to None.
+    #[cfg_attr(not(test), expect(dead_code))]
+    fn next(&mut self) -> Result<Option<&NodeState>, FileIoError> {
+        firewood_counter!(
+            "firewood.change_proof.next",
+            "number of next calls on PreOrderIterator",
```

Comment:
nobody will know what PreOrderIterator is in this context
```suggestion
            "number of nodes iterated to calculate a change proof",
```